test/e2e/test/beat/pod_builder.go

```diff
@@ -0,0 +1,238 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package beat
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/test/e2e/cmd/run"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test"
+)
+
+// Builder to create Beats
+type PodBuilder struct {
+	Pod corev1.Pod
+}
+
+func NewPodBuilder(name string) PodBuilder {
+	meta := metav1.ObjectMeta{
+		Name:      name,
+		Namespace: test.Ctx().ManagedNamespace(0),
+		Labels:    map[string]string{run.TestNameLabel: name},
+	}
+
+	return PodBuilder{
+		Pod: corev1.Pod{
+			ObjectMeta: meta,
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{
+					{
+						Name:  "ubuntu",
+						Image: "ubuntu",
+						Command: []string{
+							"bash",
+							"-c",
+							"while [ true ]; do echo \"$(date)\"; sleep 5; done",
+						},
+					},
+				},
+				SecurityContext: test.DefaultSecurityContext(),
+			},
+		},
+	}
+}
+
+func (pb PodBuilder) RuntimeObjects() []runtime.Object {
+	return []runtime.Object{&pb.Pod}
+}
+
+func (pb PodBuilder) InitTestSteps(k *test.K8sClient) test.StepList {
```

Comment:
shouldn't all those test steps be attached to the Beat builder? I'm not sure I understand the PodBuilder concept.

Reply:
PodBuilder is the builder that can be used to build a Pod that will be deployed during the test. Since for some test scenarios we need an actual Pod doing something to validate our Beat deployment.

Not all Beat tests will require a pod (e.g. heartbeat test doesn't) and we may need a different behavior of this pod depending on the test. It may need to be deployed to a different namespace, with different labels, we might need more than a single pod, it might need to log in particular way or expose an endpoint. I'd rather not overload Beat Builder with all of those. I think we have a nice (-ish) abstraction for creating resources and PodBuilder uses it.